bulk: use expiration time to disable automatic merging in import, backfill, restore

jeffrey-xiao opened the pull request:
The existing mechanism to prevent the merge queue from automatically merging splits created in import, backfill, and restore was to gossip the table keys that the merge queue should ignore when scanning
replicas. Now that there is support for specifying an expiration time at a range level, we can use that instead of the gossip mechanism.

All splits created during backfill, restore, and import use an expiration time of an hour. The rationale behind using an expiration time rather than unsplitting the ranges at the end of the operation is because adding an additional O(n) cost to bulk operations is not ideal when n can be large.

Fixes #37697.

@dt I don't have context on how long import jobs take, so an hour might be too conservative.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/38079)
<!-- Reviewable:end -->

nvb reviewed (changes requested):
The code here looks mostly complete, but it's unclear to me whether it is implementing the exact approach we want. There were a few ideas floated around in #37697, so I understand why there might be some confusion. I'm also a bit confused about what we agreed upon, so input from @dt would be appreciated.

Also, after this change, it looks like `SchemaChanger.backfillIndexes` will be the last remaining users of `Gossip.DisableMerges` (other than the pre-migration paths). It would be nice to make a similar change there in a different PR so that we can eventually go through and remove the code entirely.

The other big point I'd leave here is that there are a decent number of new conditions introduced in this PR. It's critical that this kind of stuff gets well commented when introduced or it becomes significantly more difficult to understand out of context when working on tangentially related changes. It's usually easy enough to parse what a block of code is doing, but in a large code base, it can become harder to understand why a block of code is doing what it's doing. For instance, why does condition X allow us to skip operation Y? Why would we ever want to perform operation Y? Leaving descriptive comments about this can save a lot of time in the future.

> Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-:-Lh_t3sfBOrV7RPdsqYJ:bhd8q5e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @jeffrey-xiao)

___
*[pkg/ccl/backupccl/restore.go, line 1079 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_saMv2fED-2FYPQGT:-Lh_saMv2fED-2FYPQGU:bsf13xp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/backupccl/restore.go#L1079)):*
> ```Go
> 	}
> 
> 	// TODO(jeffreyxiao): Remove this check in 20.1.
> ```

Explain this in a comment. Why do we only disable merging if the cluster version is above this level?
___
*[pkg/ccl/backupccl/restore.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_s8Wi7X7TsNV_yokQ:-Lh_s8Wi7X7TsNV_yokR:b-vebrcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/backupccl/restore.go#L1255)):*
> ```Go
> 	}
> 
> 	// Unsplit ranges after restoration is complete.
> ```

In https://github.com/cockroachdb/cockroach/issues/37697#issuecomment-497035158 it sounded like we were moving towards an approach where we would never manually unsplit the ranges after the restore. Instead, we would set a reasonable TTL on the split points to give the restore enough time to start importing into the split ranges and simply let the split points expire. Where did we land on that? What would be a reasonable TTL? cc. @dt 

EDIT: I see below that we do something like this for IMPORT. Why is RESTORE different?
___
*[pkg/ccl/importccl/import_stmt.go, line 786 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_sfvDCUpg1FbFgqEe:-Lh_sfvDCUpg1FbFgqEf:b-nqqoba) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/importccl/import_stmt.go#L786)):*
> ```Go
> 	}
> 
> 	stickyBitEnabled := r.settings.Version.IsActive(cluster.VersionStickyBit)
> ```

Same thing here. Give these descriptive comments about why we're only doing this when this version is inactive.
___
*[pkg/sql/backfill.go, line 1101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_t-Th-_-wHQrPgi33:-Lh_t-Th-_-wHQrPgi34:b-a02hno) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/sql/backfill.go#L1101)):*
> ```Go
> 		return err
> 	}
> 	// Unsplit ranges after backfill and validation are complete.
> ```

Same point as above. Is this still the approach we want?
___
*[pkg/storage/replica_command.go, line 384 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_rDclDP-pH97aOCIy:-Lh_rDclDP-pH97aOCIz:b-a81j3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d557c8fd3596acb804e3e883fec0b8008718d88/pkg/storage/replica_command.go#L384)):*
> ```Go
> ) (roachpb.AdminUnsplitResponse, *roachpb.Error) {
> 	var reply roachpb.AdminUnsplitResponse
> 	var lastErr error
> ```

Instead of copying all of this from `AdminSplit`, can we pull out a utility function? At some point, we might need the same logic structure for `AdminMerge` as well.
___
*[pkg/storage/replica_command.go, line 469 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_rcKR-fdeXOnz0gG8:-Lh_rcKR-fdeXOnz0gG9:b5aoteb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d557c8fd3596acb804e3e883fec0b8008718d88/pkg/storage/replica_command.go#L469)):*
> ```Go
> 		return txn.Run(ctx, b)
> 	})
> 	// The ConditionFailedError can occur because the descriptors acting as
> ```

We might as well mirror the same logic as that in `adminSplitWithDescriptor` and put this inside of a `; err != nil {` block.


<!-- Sent from Reviewable.io -->

jeffrey-xiao committed:
storage: add retry logic to AdminUnsplit

This commit extracts the retry logic in AdminSplit into a helper
function and uses for AdminUnsplit.

Release note: None

jeffrey-xiao committed:
bulk: use expiration time to disable automatic merging in import, backfill, restore

The existing mechanism to prevent the merge queue from automatically
merging splits created in import, backfill, and restore was to gossip
the table keys that the merge queue should ignore when scanning
replicas. Now that there is support for specifying an expiration time at
a range level, we can use that instead of the gossip mechanism.

All splits created during backfill, restore, and import use an
expiration time of an hour. The rationale behind using an expiration
time rather than unsplitting the ranges at the end of the operation is
because adding an additional O(n) cost to bulk operations is not ideal
when n can be large.

Release note: None

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/ccl/backupccl/restore.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_s8Wi7X7TsNV_yokQ:-Lhaa7aq30ZDLZUXtcmT:bemvmkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/backupccl/restore.go#L1255)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In https://github.com/cockroachdb/cockroach/issues/37697#issuecomment-497035158 it sounded like we were moving towards an approach where we would never manually unsplit the ranges after the restore. Instead, we would set a reasonable TTL on the split points to give the restore enough time to start importing into the split ranges and simply let the split points expire. Where did we land on that? What would be a reasonable TTL? cc. @dt 

EDIT: I see below that we do something like this for IMPORT. Why is RESTORE different?
</blockquote></details>

Yeah, I was also under the impression that TTL'ed splits would mean we didn't need an unsplit step (and that was a big reason I was in favor of TTL'ed splits -- I didn't want to add another O(n) step to bulk ops where n can be pretty big).
___
*[pkg/ccl/importccl/sst_writer_proc.go, line 155 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-LhadbbI53rmNjpf6D5m:-LhadbbI53rmNjpf6D5n:b-1m2ik4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88dbc837f1590e3200453115a229459a430ce99/pkg/ccl/importccl/sst_writer_proc.go#L155)):*
> ```Go
> 		expirationTime := hlc.Timestamp{}
> 		if stickyBitEnabled {
> 			expirationTime = sp.db.Clock().Now().Add(time.Hour.Nanoseconds(), 0)
> ```

what harm does setting this do on older versions?
___
*[pkg/sql/backfill.go, line 1101 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_t-Th-_-wHQrPgi33:-LhacOKg4xMMG_7faaLr:bu8bbt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/sql/backfill.go#L1101)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same point as above. Is this still the approach we want?
</blockquote></details>

actually, in this case, these are not even the right spans at all -- IIRC these are the *PK spans* that we *read* from in order to generate and write index spans -- the splits, we make (well, we don't yet but we should) would be in the index, not these, and we don't have a list of them (so it is like direct IMPORT).


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @nvanbenschoten)

___
*[pkg/ccl/backupccl/restore.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_s8Wi7X7TsNV_yokQ:-LhalgfP-NqlwiUhIeAp:b-qr9x4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/backupccl/restore.go#L1255)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Yeah, I was also under the impression that TTL'ed splits would mean we didn't need an unsplit step (and that was a big reason I was in favor of TTL'ed splits -- I didn't want to add another O(n) step to bulk ops where n can be pretty big).
</blockquote></details>

So should we just a TTL on the splits and not unsplit them at the end of the operation? What should the TTL be?
___
*[pkg/ccl/importccl/sst_writer_proc.go, line 155 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-LhadbbI53rmNjpf6D5m:-LhalrmaE6vemVyQnFWd:b-e2hnkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e88dbc837f1590e3200453115a229459a430ce99/pkg/ccl/importccl/sst_writer_proc.go#L155)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

what harm does setting this do on older versions?
</blockquote></details>

On older versions `stickyBitEnabled` would be false, so the split is always considered for automatic merging if the merge queue is enabled. The behavior should be consistent in a mixed version cluster.


<!-- Sent from Reviewable.io -->

dt reviewed (approved):
bulk-io parts LGTM (leaving core parts to core reviewers)

> Reviewed 3 of 5 files at r5, 3 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-:-LhkOrUS3zeDYjoXRNpr:bwq08zw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @dt, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/ccl/backupccl/restore.go, line 1255 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lh_s8Wi7X7TsNV_yokQ:-LhkOV_QEnQaP4s85Vpn:bgewdqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/3dae6e291a59141cc54e33733a3ba877f59d4a9f/pkg/ccl/backupccl/restore.go#L1255)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

So should we just a TTL on the splits and not unsplit them at the end of the operation? What should the TTL be?
</blockquote></details>

yep, 1h is fine.


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 1 of 6 files at r3, 5 of 5 files at r5, 5 of 5 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/storage/replica_command.go, line 414 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-Lhk_21o7qmd-6m6YsnF:-Lhk_21o7qmd-6m6YsnG:blld5k6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c2bfb3e66d91a978ca38f9bcb3e7900782b2b69c/pkg/storage/replica_command.go#L414)):*
> ```Go
> }
> 
> func (r *Replica) executeCommand(ctx context.Context, command func() error) *roachpb.Error {
> ```

This is nice, but I think we can improve it a little more. To start, I'd rename it to something like `executeAdminCommandWithDescriptor`. There are plenty of other commands that would never want to call this.

We can then give it a bit of justification in a comment. Why would anyone call this? What's the point of the retry loop? To answer that, we'll end up saying something about how the function is wrapping a read-modify-write operation for RangeDescriptors in a retry loop. Retry loops around these kinds of operations are pretty common because the write will often perform a CAS-like operation to avoid write skew. We see this all the time when dealing with atomics.

Once we talk about that, it may become evident that a more useful/understandable signature is something like:
```
func(ctx context.Context, updateDesc func(*roachpb.RangeDescriptor) error) *roachpb.Error {
```


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/38079#-:-Lhl2QYIDiUM4rE1N-J_:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @dt, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
TFTRs!

bors r+
